### PR TITLE
[endpoint-test] Support Docker Compose

### PR DESCRIPTION
**Pull request by jaylees14**
Adds support for Docker compose in `temple test` - does some validation that all of the services are up before continuing

### CODE REVIEW

File: src/main/scala/temple/test/ProjectTester.scala [hunk 1]
```diff
@@ -17,27 +19,56 @@ object ProjectTester {
...
+      case Provider.Kubernetes =>
+        println("🍪 Spinning up Kubernetes infrastructure...")
+        exec(s"sh $generatedPath/deploy.sh")
+
+        // Grab Kong's URL for future requests
+        val Array(baseURL, kongAdmin, _*) = exec("minikube service kong --url").split("\n")
+        val baseIP                        = baseURL.replaceAll("http[s]*://", "")
+        ProjectConfig(baseIP, kongAdmin)
+      case Provider.DockerCompose =>
+        println("🐳 Spinning up Docker Compose infrastructure...")
```
Comment (Lewiky) [inline src/main/scala/temple/test/ProjectTester.scala:68]: Nice

File: src/main/scala/temple/test/ProjectTester.scala [hunk 2]
```diff
@@ -17,27 +19,56 @@ object ProjectTester {
...
-    val Array(baseURL, kongAdmin, _*) = exec("minikube service kong --url").split("\n")
-    val baseIP                        = baseURL.replaceAll("http[s]*://", "")
-    val config                        = ProjectConfig(baseIP, kongAdmin)
+    val config = provider match {
+      case Provider.Kubernetes =>
+        println("🍪 Spinning up Kubernetes infrastructure...")
+        exec(s"sh $generatedPath/deploy.sh")
+
+        // Grab Kong's URL for future requests
+        val Array(baseURL, kongAdmin, _*) = exec("minikube service kong --url").split("\n")
```
Comment (Lewiky) [inline src/main/scala/temple/test/ProjectTester.scala:64]: Also nice syntax, didn't know you could do that, very good

**Review by Lewiky — APPROVED** [review]
Love it, very nice

File: src/main/scala/temple/test/ProjectTester.scala [hunk 3]
```diff
@@ -17,27 +19,56 @@ object ProjectTester {
 
   /** Configure the infrastructure for testing */
   private def performSetup(templefile: Templefile, generatedPath: String): ProjectConfig = {
-    println("🍪 Spinning up Kubernetes infrastructure...")
-    // Deploy Kubernetes
-    exec(s"sh $generatedPath/deploy.sh")
+    val provider = templefile
+      .lookupMetadata[Metadata.Provider]
+      .getOrElse(throw new RuntimeException("Could not find deployment information"))
```
Comment (c-harding) [inline src/main/scala/temple/test/ProjectTester.scala]: Personally I’d prefer braces here to show that this doesn’t get immediately executed

File: src/main/scala/temple/test/ProjectTester.scala [hunk 4]
```diff
@@ -17,27 +19,56 @@ object ProjectTester {
...
+        ProjectConfig(baseIP, kongAdmin)
+      case Provider.DockerCompose =>
+        println("🐳 Spinning up Docker Compose infrastructure...")
+        exec(
+          s"cd $generatedPath && docker ps -a -q | xargs docker rm -f && docker volume prune -f && docker-compose up --build -d",
+        )
+
+        var finishedStarting = false
+        while (!finishedStarting) {
+          val kongRequest = Try(Http("http://localhost:8001/status").asString)
```
Comment (xsanda) [inline src/main/scala/temple/test/ProjectTester.scala:43]: Is it safe to use Try? Could there be any other exceptions that we don’t want to discard?

**Review by c-harding — APPROVED** [review]
Nice, though Try feels a little like cheating

File: src/main/scala/temple/test/ProjectTester.scala [hunk 5]
```diff
@@ -15,29 +19,80 @@ object ProjectTester {
...
+          if (templefile.usesAuth) {
+            // If the service uses auth, wait for that service to respond
+            val authRequest = Try(Http("http://localhost:1024/api/auth/login").asString)
+            finishedStarting = authRequest.fold(_ => false, { _.code == HttpURLConnection.HTTP_NOT_FOUND })
+          } else {
+            // Otherwise wait for kong to be available
+            val kongRequest = Try(Http("http://localhost:8001/status").asString)
+            finishedStarting = kongRequest.fold(_ => false, { _.code == HttpURLConnection.HTTP_OK })
+          }
+          exec("sleep 5")
```
Comment (c-harding) [inline src/main/scala/temple/test/ProjectTester.scala]: do you really want to sleep after it’s finished starting?

**Pull request merged**